### ISSUE TITLE
PANIC at arc.c:2524:arc_evict_state_impl() VERIFY(!(((*({ __attribute__((unused)) typeof((&((hash_lock)->m_mutex))->owner) __var = ( typeof((&((hash_lock)->m_mutex))->owner)) 0; (volatile typeof((&((hash_lock)->m_mutex))->owner) *)&((&((hash_lock)->m_mutex))->owner); }))) == get_current())) failed

### ISSUE THREAD

**Issue by JuliaVixen** [zfsonlinux/zfs#4995]
Well this happened... I was doing a zfs send|recv thing at the time. 

```
[312631.332008] VERIFY(!(((*({ __attribute__((unused)) typeof((&((hash_lock)->m_mutex))->owner) __var = ( typeof((&((hash_lock)->m_mutex))->owner)) 0; (volatile typeof((&((hash_lock)->m_mutex))->owner) *)&((&((hash_lock)->m_mutex))->owner); }))) == get_current())) failed
[312631.332120] PANIC at arc.c:2524:arc_evict_state_impl()
[312631.332174] Showing stack for process 20974
[312631.332177] CPU: 0 PID: 20974 Comm: send_traverse Tainted: P           O    4.4.6-gentoo-debug2 #1
[312631.332177] Hardware name: Supermicro X10SLL-F/X10SLL-SF/X10SLL-F/X10SLL-SF, BIOS 1.0a 06/11/2013
[312631.332178]  0000000000000000 ffff880695c62eb8 ffffffff8130fa9d ffffffffa129331e
[312631.332180]  ffffffffa12932ff ffff880695c62ec8 ffffffffa0a3a5ac ffff880695c63058
[312631.332181]  ffffffffa0a3a764 ffffc90186087000 ffff880800000028 ffff880695c63068
[312631.332183] Call Trace:
[312631.332187]  [<ffffffff8130fa9d>] dump_stack+0x4d/0x63
[312631.332193]  [<ffffffffa0a3a5ac>] spl_dumpstack+0x3d/0x3f [spl]
[312631.332196]  [<ffffffffa0a3a764>] spl_panic+0xa9/0xdc [spl]
[312631.332213]  [<ffffffffa1171067>] arc_evict_state+0x326/0x95e [zfs]
[312631.332221]  [<ffffffffa1171723>] arc_adjust_impl.constprop.25+0x2a/0x31 [zfs]
[312631.332230]  [<ffffffffa1171a84>] arc_adjust+0x35a/0x4b8 [zfs]
[312631.332238]  [<ffffffffa1173a5f>] arc_shrink+0x10d/0x10f [zfs]
[312631.332246]  [<ffffffffa1173ae9>] __arc_shrinker_func.isra.24+0x88/0xf6 [zfs]
[312631.332255]  [<ffffffffa1173b63>] arc_shrinker_func_scan_objects+0xc/0x1d [zfs]
[312631.332257]  [<ffffffff810c5962>] shrink_slab.part.56.constprop.61+0x16b/0x1fc
[312631.332258]  [<ffffffff810c787e>] shrink_zone+0x6e/0x144
[312631.332260]  [<ffffffff810c8173>] do_try_to_free_pages+0x1b3/0x2bc
[312631.332261]  [<ffffffff810c8380>] try_to_free_pages+0x89/0x90
[312631.332263]  [<ffffffff810c000f>] __alloc_pages_nodemask+0x57c/0x8d4
[312631.332265]  [<ffffffff814cecd4>] ? __schedule+0x5fe/0x744
[312631.332267]  [<ffffffff810e9b58>] cache_alloc_refill+0x26d/0x4a3
[312631.332269]  [<ffffffff810e9de6>] kmem_cache_alloc+0x58/0xae
[312631.332271]  [<ffffffffa0a36184>] spl_kmem_cache_alloc+0x10c/0xa9f [spl]
[312631.332279]  [<ffffffffa116dafe>] ? buf_cons+0x3f/0x45 [zfs]
[312631.332287]  [<ffffffffa116dabf>] ? arc_space_consume+0xa3/0xa3 [zfs]
[312631.332300]  [<ffffffffa123127a>] zio_buf_alloc+0x53/0x57 [zfs]
[312631.332308]  [<ffffffffa116def8>] arc_get_data_buf+0x33a/0x461 [zfs]
[312631.332316]  [<ffffffffa116e0e4>] arc_buf_clone+0xc5/0xff [zfs]
[312631.332325]  [<ffffffffa117362e>] ? arc_read+0xc9a/0xc9a [zfs]
[312631.332333]  [<ffffffffa1172cd7>] arc_read+0x343/0xc9a [zfs]
[312631.332335]  [<ffffffff814d0a10>] ? mutex_unlock+0x9/0xb
[312631.332346]  [<ffffffffa1196380>] traverse_visitbp+0x48a/0x762 [zfs]
[312631.332358]  [<ffffffffa11962d3>] traverse_visitbp+0x3dd/0x762 [zfs]
[312631.332369]  [<ffffffffa11962d3>] traverse_visitbp+0x3dd/0x762 [zfs]
[312631.332379]  [<ffffffffa11962d3>] traverse_visitbp+0x3dd/0x762 [zfs]
[312631.332390]  [<ffffffffa11962d3>] traverse_visitbp+0x3dd/0x762 [zfs]
[312631.332400]  [<ffffffffa11962d3>] traverse_visitbp+0x3dd/0x762 [zfs]
[312631.332411]  [<ffffffffa11962d3>] traverse_visitbp+0x3dd/0x762 [zfs]
[312631.332421]  [<ffffffffa1196b42>] traverse_dnode+0xb3/0x161 [zfs]
[312631.332431]  [<ffffffffa1196516>] traverse_visitbp+0x620/0x762 [zfs]
[312631.332434]  [<ffffffffa0a37a8e>] ? taskq_dispatch+0x22c/0x23e [spl]
[312631.332444]  [<ffffffffa1196990>] traverse_impl+0x338/0x437 [zfs]
[312631.332446]  [<ffffffff8106176d>] ? enqueue_task_fair+0xb83/0xbe3
[312631.332456]  [<ffffffffa119200d>] ? dump_bytes_cb+0x123/0x123 [zfs]
[312631.332467]  [<ffffffffa1196deb>] traverse_dataset_resume+0x4d/0x4f [zfs]
[312631.332477]  [<ffffffffa1192175>] ? dmu_recv_existing_end+0x6f/0x6f [zfs]
[312631.332487]  [<ffffffffa1192045>] send_traverse_thread+0x38/0x75 [zfs]
[312631.332490]  [<ffffffffa0a36fa1>] thread_generic_wrapper+0x9d/0xaa [spl]
[312631.332491]  [<ffffffffa0a36f04>] ? __thread_create+0x13f/0x13f [spl]
[312631.332493]  [<ffffffff8105533a>] kthread+0xcd/0xd5
[312631.332495]  [<ffffffff8105526d>] ? kthread_freezable_should_stop+0x43/0x43
[312631.332496]  [<ffffffff814d265f>] ret_from_fork+0x3f/0x70
[312631.332497]  [<ffffffff8105526d>] ? kthread_freezable_should_stop+0x43/0x43
```

`root     20974  0.0  0.0      0     0 ?        DN   Aug20   0:01 [send_traverse]`

This is the "zfs-master" from Aug 12, 2016; built with ` ./configure --enable-debug --enable-debug-kmem`

```
Aug 17 05:48:42 localhost kernel: SPL: Loaded module v0.6.5-1 (DEBUG mode)
Aug 17 05:48:42 localhost kernel: ZFS: Loaded module v0.6.5-1 (DEBUG mode), ZFS pool version 5000, ZFS filesystem version 5
```

ECC memory and all that... I think I was running this at the time..

`zfs send -eLv -R r@Aug_18_2016 | zfs recv -evsF Q`

Oh, I should mention, the zpool named "r" here is a mirror-0 with two disks, and earlier, one of the disks was disconnected while in use. Everything was fine at the time, and a for while afterwards. The pool was mounted "-o readonly=on".

At some point before the above kernel panic, I sent a `kill -19` signal to the "zfs send" process, and then sent a `kill -18` signal, because I needed to temporarily pause it for (reasons). It seemed to be ok, then I came back later and this kernel panic had happened.

**Comment by dweeezil** [comment]
Same as #4912.  Apparently we didn't have a PR to fix this.  See the just-posted #4998.

**Comment by behlendorf** [comment]
Closing, resolved by #4998.
